### PR TITLE
feat(rust): added the possibility to overwrite http headers in inlets

### PR DESCRIPTION
**Pull request by davide-baldo**
This PR generalizes the HTTP interception code used in InfluxDb, and adds the possibility to write custom http headers using `ockam tcp-inlet create --http-header "Header: value" ...`

### CODE REVIEW

File: implementations/rust/ockam/ockam_api/src/http/state.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+pub(crate) trait ClientRequestWriter {
+    fn write_headers(
+        &self,
+        request: &httparse::Request,
+        buffer: &mut Vec<u8>,
+    ) -> ockam_core::Result<()>;
+}
+
+impl RequestState {
+    /* Parse the incoming data, attaching an Authorization header token to it.
```
Comment (etorreborre) [inline implementations/rust/ockam/ockam_api/src/http/state.rs]: This should use `///` for commenting.

File: implementations/rust/ockam/ockam_command/src/tcp/inlet/create.rs [hunk 2]
```diff
@@ -186,6 +195,43 @@ impl Command for CreateCommand {
...
+                        &overwrite_http_header_address,
+                        cmd.http_header.clone(),
+                    )
+                    .await;
+
+                match result {
+                    Ok(_) => {
+                        if let Some(pb) = pb.as_ref() {
+                            let created_message = format!(
+                                "Created a new Interceptor Service bound to {}\n",
```
Comment (etorreborre) [inline implementations/rust/ockam/ockam_command/src/tcp/inlet/create.rs]: "a new HTTP Interceptor Service"

File: implementations/rust/ockam/ockam_command/src/util/parsers.rs [hunk 3]
```diff
@@ -21,6 +21,17 @@ pub(crate) fn hostname_parser(input: &str) -> Result<SchemeHostnamePort> {
...
 
+/// Helper function for parsing a key-value pair from user input
+/// The input is expected to be in the format `key:value`
+pub(crate) fn http_header_parser(input: &str) -> Result<(String, String)> {
+    let parts: Vec<&str> = input.split(':').collect();
+    if parts.len() != 2 {
+        return Err(miette!("Invalid header format. Expected 'key:value'"));
+    }
+
+    Ok((parts[0].trim().to_string(), parts[1].trim().to_string()))
```
Comment (etorreborre) [inline implementations/rust/ockam/ockam_command/src/util/parsers.rs]: Ideally we could check [the specification for header names and values](https://datatracker.ietf.org/doc/html/rfc7230#page-22), which is like:

- The header name has one or more characters in this range:
`! # $ % & ' * + - .  ^ _ ` | ~ digit alpha`
```
- The value is non-empty, and does not contain newlines (`\n`, `\r\n`) or control characters (`0x00–0x1F, 0x7F`).

This way we can prevent hard-to-diagnose issues down the stream. One typical example would be to define:
```
--header token:$AUTH_TOKEN
```
Where the user makes a mistake and uses `AUTH_TOKEN` instead of `AUTHORIZATION_TOKEN`.

File: implementations/rust/ockam/ockam_api/src/http/interceptor.rs [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+};
+use std::io::Write;
+use std::sync::{Arc, Mutex as SyncMutex};
+
+struct HttpHeadersInterceptorFactory {
+    headers: Arc<Vec<(String, String)>>,
+}
+
+impl HttpHeadersInterceptorFactory {
+    pub async fn create(
```
Comment (etorreborre) [inline implementations/rust/ockam/ockam_api/src/http/interceptor.rs]: The naming of these functions make things a bit hard to follow IMO.

 1. `HttpHeadersInterceptorFactory::create` creates something but we don't know what.
 2. Internally, it calls `PortalInletInterceptor::create` which also creates something but does not return it, so we don't really what it is.
 3. But then we see that to create a `PortalInletInterceptor` we need a `HttpHeadersInterceptorFactory`, so presumably step 1. creates more than just a `HttpHeadersInterceptorFactory`.
 4. We then conclude that a `PortalInletInterceptor` uses a factory which also creates an `Interceptor`, so we eventually have interceptors creating interceptors 🙃.

It would be nice to have a clearer terminology:

- With maybe `register/start` instead of `create` when there are side-effects.
- Some documentation for methods having side-effects so we don't have to look at the implementation to understand what's going on.
- Just one level of interception and not an interceptor that can create other interceptors.

**Review by etorreborre — DISMISSED** [review]
🏯

File: implementations/rust/ockam/ockam_command/src/util/parsers.rs (hunk 3 above)
Comment (davide-baldo) [reply implementations/rust/ockam/ockam_command/src/util/parsers.rs]: I've taken inspiration from the `httparse` library we're using and made the parsing much more robust!
And I've disallowed an empty value, so we can cover the empty variable resolution case.

File: implementations/rust/ockam/ockam_api/src/http/interceptor.rs (hunk 4 above)
Comment (davide-baldo) [reply implementations/rust/ockam/ockam_api/src/http/interceptor.rs]: I've added some comment and renamed create into `start_listener`, which should be much more self-explanatory.
That said, I agree that the situation is still convoluted.
Oleksandr did propose to move the interceptors from the communication protocol and place them directly adjacent to the inlet and output ports. This would drastically simplify the architecture and avoid extra decoding/encoding of the payload.

**Review by etorreborre — APPROVED** [review]
